server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/GroupsRoutesTest.java

```diff
@@ -0,0 +1,465 @@
+import static org.apache.james.webadmin.Constants.SEPARATOR;
+import static org.apache.james.webadmin.WebAdminServer.NO_CONFIGURATION;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.hamcrest.CoreMatchers.is;
+import static org.mockito.Matchers.anyString;
+import static org.mockito.Mockito.doThrow;
+import static org.mockito.Mockito.mock;
+
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+
+import org.apache.james.metrics.logger.DefaultMetricFactory;
+import org.apache.james.rrt.api.RecipientRewriteTable;
+import org.apache.james.rrt.api.RecipientRewriteTableException;
+import org.apache.james.rrt.memory.MemoryRecipientRewriteTable;
+import org.apache.james.webadmin.WebAdminServer;
+import org.apache.james.webadmin.utils.JsonTransformer;
+import org.eclipse.jetty.http.HttpStatus;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import com.jayway.restassured.RestAssured;
+import com.jayway.restassured.builder.RequestSpecBuilder;
+import com.jayway.restassured.filter.log.LogDetail;
+import com.jayway.restassured.http.ContentType;
+
+import de.bechte.junit.runners.context.HierarchicalContextRunner;
+
+@RunWith(HierarchicalContextRunner.class)
+public class GroupsRoutesTest {
+
+    private static final String DOMAIN = "b.com";
+    private static final String GROUP1 = "group1" + "@" + DOMAIN;
+    private static final String GROUP2 = "group2" + "@" + DOMAIN;
+    private static final String USER_A = "a" + "@" + DOMAIN;
+    private static final String USER_B = "b" + "@" + DOMAIN;
+
+    private WebAdminServer webAdminServer;
+
+    private void createServer(GroupsRoutes groupsRoutes) throws Exception {
+        webAdminServer = new WebAdminServer(
+            new DefaultMetricFactory(),
+            groupsRoutes);
+        webAdminServer.configure(NO_CONFIGURATION);
+        webAdminServer.await();
+
+        RestAssured.requestSpecification = new RequestSpecBuilder()
+            .setContentType(ContentType.JSON)
+            .setAccept(ContentType.JSON)
+            .setConfig(newConfig().encoderConfig(encoderConfig().defaultContentCharset(StandardCharsets.UTF_8)))
+            .setPort(webAdminServer.getPort().toInt())
+            .setBasePath(GroupsRoutes.ROOT_PATH)
+            .log(LogDetail.ALL)
+            .build();
+    }
+
+    @After
+    public void stop() {
+        webAdminServer.destroy();
+    }
+
+    public class NormalBehaviour {
+
+        @Before
+        public void setUp() throws Exception {
+            MemoryRecipientRewriteTable memoryRecipientRewriteTable = new MemoryRecipientRewriteTable();
+            memoryRecipientRewriteTable.addErrorMapping("error", DOMAIN, "disabled");
+            memoryRecipientRewriteTable.addRegexMapping("regex", DOMAIN, ".*@b\\.com");
+            memoryRecipientRewriteTable.addAliasDomainMapping("alias", DOMAIN);
```

Comment:
you never use these mappings. I understand that you put them to check that they does not interfere, but maybe you should create them in an other test case, so that you have: NormalBehaviour, ErrorBehaviour, NormalBehaviourWithAlreadyDefinedRules

Reply:
how would you do it ? Basically, I would like all the tests from NormalBehaviour to be replayed in NormalBehaviourWithAlreadyDefinedRules

Reply:
I know you don't like it, but the simpler way is to have one class extending the other one, or nicer having them extending a common abstract class

Reply:
ugly inheritance done